sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs

```diff
@@ -198,43 +236,37 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
 
             try
             {
-                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception e)
-            {
-                scope.Failed(e);
-                throw;
-            }
-        }
-
-        protected async ValueTask<Response> ApplyStateAsync(bool async, Response response, bool hasCompleted, bool hasSucceeded, RequestFailedException? requestFailedException, bool throwIfFailed = true)
-        {
-            RawResponse = response;
+                var state = await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+                RawResponse = state.RawResponse;
+                if (!state.HasCompleted)
+                {
+                    return state.RawResponse;
+                }
 
-            if (!hasCompleted)
-            {
-                return response;
-            }
+                if (!state.HasSucceeded && state.OperationFailedException == null)
+                {
+                    var exception = async
+                        ? await _diagnostics.CreateRequestFailedExceptionAsync(state.RawResponse).ConfigureAwait(false)
+                        : _diagnostics.CreateRequestFailedException(state.RawResponse);
+                    state = OperationState.Failure(state.RawResponse, exception);
+                }
 
-            HasCompleted = true;
-            if (hasSucceeded)
-            {
-                return response;
-            }
+                asyncLock.SetValue(state);
 
-            OperationFailedException = requestFailedException ??
-                (async
-                    ? await _diagnostics.CreateRequestFailedExceptionAsync(response).ConfigureAwait(false)
-                    : _diagnostics.CreateRequestFailedException(response));
+                if (state.OperationFailedException != null)
+                {
+                    throw state.OperationFailedException;
+                }
 
-            if (throwIfFailed)
+                return state.RawResponse;
+            }
+            catch (Exception e)
             {
-                throw OperationFailedException;
+                scope.Failed(e);
+                throw;
             }
-
-            return response;
         }
 
-        protected abstract ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken);
+        protected abstract ValueTask<OperationState> UpdateStateAsync(bool async, CancellationToken cancellationToken);
```

Comment:
It feels weird that we have a `State` class representing the state of the current operation yet that state is never stored in the instance of the class it simply creates one, then assigns some member variables.  Should we maybe store this as a member variable and have the HasCompleted / HasValue / Value all lambda into this object to get their answers?

Reply:
It is stored: https://github.com/Azure/azure-sdk-for-net/pull/27639/files#diff-f9f7a9d046f21930f94a8ca1d25df78506aabb692ce00a3504bdce32498b4ef3R254